icu4c/source/test/intltest/utfiteratortest.cpp

```diff
@@ -487,8 +487,23 @@ class UTFIteratorTest : public IntlTest {
         static_assert(!std::ranges::common_range<CodeUnitRange>);
         static_assert(std::ranges::input_range<CodeUnitRange>);
         static_assert(!std::ranges::forward_range<CodeUnitRange>);
-        auto it = utfIterator<char32_t, UTF_BEHAVIOR_FFFD>(codeUnits.begin(), codeUnits.end());
-        // TODO: Use `it`.
+        // TODO(egg): UTFStringCodePoints could support this.
+        struct CodePoints : std::ranges::view_interface<CodePoints> {
+            using iterator =
+                UTFIterator<char32_t, UTF_BEHAVIOR_FFFD, std::ranges::iterator_t<CodeUnitRange>,
+                            std::ranges::sentinel_t<CodeUnitRange>>;
+            CodePoints(CodeUnitRange codeUnits) : codeUnits(codeUnits) {}
+            iterator begin() { return iterator(codeUnits.begin(), codeUnits.end()); }
+            std::default_sentinel_t end() {
+                return std::default_sentinel;
+            }
```

Comment:
Why not something like
```
auto end() { return codeUnits.end(); }
```
?

Reply:
Because I wrote `bool operator==(default_sentinel_t)`, instead of `template<sentinel_for<UnitIter> Sentinel> bool operator==(Sentinel s)`.

I think since it’s an iterator that knows its end, it makes sense to make the default sentinel a sentinel for it, see https://eel.is/c++draft/default.sentinel.

For the one that does not know its end, we should have `template<sentinel_for<UnitIter> Sentinel> bool operator==(Sentinel s)`, and there the code point range should return the sentinel of the underlying code unit range.

Reply:
I am trying to make it work with C++17 as well, see my attempt in the generic (non-input-iterator) UTFIterator.

I assume that if someone uses any kind of sentinel type, then I can compare the current UnitIter with the sentinel object. This should work with both validating and unsafe code point iterators which would be very nice.

I also found bugs with using the right kind of UTFImpl, and your operator==(sentinel) is missing a condition which should make it not work correctly.

Reply:
> I assume that if someone uses any kind of sentinel type, then I can compare the current UnitIter with the sentinel object.

Yes.

But in addition to that, our validating iterators (and not the unsafe ones) are iterators that know their bound, so it seems nice to make the default sentinel a sentinel for them, regardless of whether the underlying code unit iterator has a separate sentinel.

Effectively, this is giving them a .hasNext(), except that it is spelled != std::default_sentinel.

Reply:
> your operator==(sentinel) is missing a condition which should make it not work correctly.

Ah yes, it should be incorrect if it is used after *. Coming up with a test that actually exercises that seems tricky though; normally you do the == after the ++ (to see that you can *), _then_ you *.